### ISSUE TITLE
Update to tendermint 0.25 (or 0.26)

### ISSUE THREAD

**Issue by ethanfrey** [iov-one/weave#133]
0.22 seems to be a bit unstable... still clarifying the merge. We wait for 0.24 (released) or 0.25 ("final" tendermint testnet)

This merges everything into tendermint, and if we don't want to import it all into our codebase, we need to do the following:

- [ ] Pull out abci repo with current protobuf definitions
- [ ] Fork iavl repo and cleanup
- [ ] Fork tmlog repo (or just depend directly on go-kit)
- [ ] Update code to newest abci standard
- [ ] Update the client to use newer go-amino rpc

Basically, all parts of the tendermint codebase that we directly import into weave/bcp-demo should be owned by us. Mainly this is forking an older version and cleaning up the dependency graph.

Ethan is comfortable with abci and iavl repos and can do any needed changes there. Maybe someone else wants to rethink the logging dependencies.

**Comment by ethanfrey** [comment]
On reflection this could be two issues:

1. is to upgrade to tendermint 0.25 the simplest way possible, which is just importing tendermint repo entirely, updating all imports, fixing any breaking api changes.
2. (multiple parts) evalulate the packages we use from tendermint and see if it is possible to extract those pieces as separate repos to depend on, or whether that will just be too much maintenance to track changes from upstream. Some pieces will have to track upstream like abci definitions. Other not, like iavl.

If we just focus on part #1 here it is not too bad and can get us a new, more stable tendermint with some useful bugfixes.

**Comment by alpe** [comment]
As discussed in slack:
💪 let's do IAVL in a separate PR with the tendermint upstream package instead of our fork

**Comment by ethanfrey** [comment]
@ruseinov 

re: your earlier questions about iavl. I just took a look and while the names changed, the functionality did not. Compare the code

Immutable tree (0.11): https://github.com/tendermint/iavl/blob/master/immutable_tree.go
is basically the same as
Tree (0.9.2): https://github.com/tendermint/iavl/blob/v0.9.2/tree.go

Mutable tree (0.11): https://github.com/tendermint/iavl/blob/master/mutable_tree.go
combined and improved:
Ophaning tree (0.9.2): https://github.com/tendermint/iavl/blob/v0.9.2/orphaning_tree.go
Versioned tree (0.9.2): https://github.com/tendermint/iavl/blob/v0.9.2/versioned_tree.go

We used Versioned tree to keep track of multiple versions. We should migrate to Mutable tree which contains the same functionality. But be careful if there are subtle changes in the return values / semantics

**Comment by ruseinov** [comment]
@alpe This is done, but if we still want to evaluate tm packages, then we should keep the ticket.

**Comment by ethanfrey** [comment]
I would create a new ticket about extracting tendermint libs (iavl, abci, any else?) as weave dependencies. Link it to this issue and close this one.

The extraction ticket is low priority for now. Once we are close to feature completion for Bns, there is lots of optimization and tuning we can do

**Comment by ethanfrey** [comment]
I would...

Meaning before anyone closes the issue, they should.

I may be slow getting around to it
